python/ray/workflow/tests/test_checkpoint.py

```diff
@@ -0,0 +1,236 @@
+import ray
+import time
+import pytest
+from ray.tests.conftest import *  # noqa
+
+import numpy as np
+from ray import workflow
+from ray.workflow.tests import utils
+from ray.workflow import storage, workflow_storage
+from ray.exceptions import RaySystemError
+
+
+@workflow.step
+def large_input():
+    return np.arange(2**24)
+
+
+@workflow.step
+def identity(x):
+    return x
+
+
+@workflow.step
+def average(x):
+    return np.mean(x)
+
+
+@workflow.step
+def checkpoint_dag(checkpoint):
+    x = large_input.options(name="large_input", checkpoint=checkpoint).step()
+    y = identity.options(name="identity", checkpoint=checkpoint).step(x)
+    return average.options(name="average").step(y)
+
+
+def _assert_step_checkpoints(wf_storage, step_id, mode):
+    result = wf_storage.inspect_step(step_id)
+    if mode == "all_skipped":
+        assert not result.output_object_valid
+        assert result.output_step_id is None
+        assert not result.args_valid
+        assert not result.func_body_valid
+        assert not result.step_options
+    elif mode == "output_skipped":
+        assert not result.output_object_valid
+        assert result.output_step_id is None
+        assert result.args_valid
+        assert result.func_body_valid
+        assert result.step_options is not None
+    elif mode == "checkpointed":
+        assert (result.output_object_valid
+                or result.output_step_id is not None)
+    else:
+        raise ValueError("Unknown mode.")
+
+
+def test_checkpoint_dag(workflow_start_regular):
+    # warm up
+    for _ in range(3):
+        outputs = checkpoint_dag.step(True).run()
+        assert np.isclose(outputs, 8388607.5)
```

Comment:
Why do we need warm up here?

Reply:
Because it prints the duration at the end of execution. This helps us understand if checkpointing works as expected. Without warmup, the duration is off by a lot. (Note: because the CI env is still flaky, so in the unit test we still check the correctness by checking the existence of checkpoint files. However showing the duration is just too helpful for manual correctness check).